orca-queue-redis/src/main/kotlin/com/netflix/spinnaker/orca/q/redis/RedisQueue.kt

```diff
@@ -58,66 +58,74 @@ class RedisQueue(
   private val messagesKey = queueName + ".messages"
   private val attemptsKey = queueName + ".attempts"
   private val locksKey = queueName + ".locks"
+  private val hashKey = queueName + ".hash"
+  private val hashesKey = queueName + ".hashes"
 
   override fun poll(callback: (Message, () -> Unit) -> Unit) {
     pool.resource.use { redis ->
-      redis.apply {
-        pop(queueKey, unackedKey, ackTimeout)
-          ?.also { id -> hincrBy(attemptsKey, id, 1) }
-          ?.let { id ->
-            readMessage(id) { payload ->
-              callback.invoke(payload) {
-                ack(id)
-                fire<MessageAcknowledged>()
-              }
-            }
+      redis.zrangeByScore(queueKey, 0.0, score(), 0, 1)
+        .firstOrNull()
+        ?.takeIf { id -> redis.acquireLock(id) }
+        ?.also { id ->
+          val ack = this::ackMessage.curry(id)
+          redis.readMessage(id) { message ->
+            callback(message, ack)
           }
```

Comment:
I think this clarifies how the callbacks work a little, at least I hope so.

`callback` is the lambda `QueueProcessor` uses to find a handler, dispatch the message (on another thread) and then call the `ack` function.